File: python/1717(유니온파인드).py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)


def find(a):
    if a == parent[a]:
        return a
    tmp = find(parent[a])
    parent[a] = tmp
    return parent[a]


def union(a, b):
    a_root = find(a)
    b_root = find(b)

    if a_root != b_root:
        parent[a_root] = b_root


n, m = map(int, input().split())
parent = [0] * (n+1)
for i in range(n+1):
    parent[i] = i


for i in range(m):
    a, b, c = map(int, input().split())

    if a == 0:
        union(b, c)
    else:
        if find(b) == find(c):
            print("YES")
        else:
            print("NO")

File: 백준/c++/1717(유니온파인드).cpp
#include <bits/stdc++.h>

using namespace std;
int parent[1000001];

int find(int x)
{
    if (parent[x] == x)
        return x;
    // 경로 압축을 위해 이렇게 표현
    return parent[x] = find(parent[x]);
}

// v를 부모로 
//   v
//   |
//   u
void union_tree(int a, int b)
{
    int aRoot = find(a);
    int bRoot = find(b);
    if (aRoot != bRoot)
        parent[aRoot] = bRoot;
}

int main(void) {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;
    int a, b, c;

    for (int i = 0; i <= n; i++)
        parent[i] = i;

    for (int i = 0; i < m; i++)
    {
        cin >> a >> b >> c;

        if (a == 0)
        {
            union_tree(b, c);
        }
        else if (a == 1)
        {
            if (find(b) == find(c))
                cout << "YES\n";
            else
                cout << "NO\n";
        }
    }
    return 0;
}